### ISSUE TITLE
LMS/Daphile Soundcloud plugin. Login blocked

### ISSUE THREAD

**Issue by DrJack99** [danielvijge/SqueezeCloud#47]
Any new policy by Soundcloud in order to LMS/daphile integration? I can't login anymore.
Thank you for your help
Roberto
![Souncloud Daphile](https://github.com/user-attachments/assets/2bf6aad1-da78-4e52-a77b-c32ffecb9db5)

**Comment by danielvijge** [comment]
That error message is produced by https://github.com/danielvijge/SqueezeCloud/blob/fbae89f5dbc558f921e77a9da2c25a8c6d1fa554/callback.html#L24 It's a piece of simple JavaScript to display the `access_token` that is present in the URL.
Could you see what your url is? Is the access token present? If so, it might be a browser issue. Any errors in the console log?

**Comment by danielvijge** [comment]
_ignore my previous (deleted) message_

Not sure if this is related to the [migration to OAuth 2.1](https://developers.soundcloud.com/blog/oauth-migration). But this should have been disabled by 1 October 2024.

This plugin is using the JavaScript SDK, which I assume would continue to work, but I now see that this is also deprecated.

Solution might be to use either a browser-based or server-side OAuth 2.1 library.
But really not sure when I can find to time to investigate.

As always, any pull request is welcome.

**Comment by Sabb0** [comment]
I have the same as OP, unfortunately. 

Do you have a slightly more ELI5 method to get this working? Thank you :)

**Comment by danielvijge** [comment]
Did a first quick investigation. It is indeed related to the new OAuth 2.1 login requirements.

A flow to solve this would be:
- When opening the settings page generate a PKCE token
- When user clicks on the Login button, open the SoundCloud login page with the PKCE token (and maybe make server name as `state` parameter so it can also be displayed on the callback page )
- After login, it redirects to the callback page. Modify this page to extract `code` from the query parameters. Display this to the user (and maybe with the server name from `state` parameter)
- Use copies the code and fills this in on a form on the settings page. User clicks on an Authorise button
- POST to the token endpoint using the code and earlier generated PKCE.
- Extract the `access_token` and `refresh_token` and store it in the settings

Additionally:
- Each API request would also need to verify if the acess token is still valid, else use the refresh token to get a new access token.

**Comment by danielvijge** [comment]
Released a first test version with the new login as 0.15.11-dev. This can be installed from the dev channel (see readme for instructions).
